Log early messages with -printtoconsole

ajtowns opened the pull request:
This allows any log messages sent prior to the `-printtoconsole` option being parsed to be replayed onto the console. This could be handy if option parsing (parameters/config file) wants to output any warnings or info messages.

promag commented:
Concept ACK.

ajtowns committed:
Replace OpenDebugLog() with StartLogging()

StartLogging() is used to mark the start of logging generically, whether
using -printtoconsole or -debuglogfile.

ajtowns committed:
Log early messages with -printtoconsole

This ensures log messages prior to StartLogging() are replayed to
the console as well as to the debug log file.

ajtowns committed:
logging.cpp: use CCriticalSection instead of std::mutex

Note: this introduces a circular dependency between logging and sync,
since sync does logging, and logging does synchronisation.

ajtowns commented:
Rebased on top of 12954 and 13148. Nitpicking welcome.

MarcoFalke commented:
Is this of any relevance? I had the impression that parameter interaction was done after logging was set up.

MarcoFalke commented:
Noting that replacing the `LogPrintf` with `FileWriteStr` fixes #13157

MarcoFalke commented:
Needs rebase again and then the circular dependency documented in the linter or the linter removed, I think.

MarcoFalke commented:
@ajtowns Are you still working on this?

ajtowns commented:
Rebased, dropped overlapping commit with #13159 (which probably should be merged first), and note that CCriticalSection change introduces dependency loop.

practicalswift commented on test/lint/lint-circular-dependencies.sh:
```diff
@@ -12,6 +12,7 @@ EXPECTED_CIRCULAR_DEPENDENCIES=(
     "chainparamsbase -> util -> chainparamsbase"
     "checkpoints -> validation -> checkpoints"
     "index/txindex -> validation -> index/txindex"
+    "logging -> sync -> logging"
```
What would be required to avoid introducing this circular dependency? :-)

MarcoFalke commented on src/logging.h:
```diff
@@ -59,9 +61,10 @@ namespace BCLog {
-        FILE* m_fileout = nullptr;
-        std::mutex m_file_mutex;
-        std::list<std::string> m_msgs_before_open;
+        CCriticalSection m_cs_log;
```
nit: 
```suggestion
        RecursiveMutex m_cs_log;
```

maflcko reviewed (commented):
Needs rebase to have travis run again

ajtowns commented:
Closing in favour of #16112
